gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/GraphManager.java

```diff
@@ -21,139 +21,101 @@
-
-    private final Map<String, Graph> graphs = new ConcurrentHashMap<>();
-    private final Map<String, TraversalSource> traversalSources = new ConcurrentHashMap<>();
+import java.util.function.Function;
 
+public interface GraphManager {
     /**
-     * Create a new instance using the {@link Settings} from Gremlin Server.
+     * @Deprecated This returns a {@link Map} that should be immutable. Please refer to
+     * getGraphNames() for replacement.
+     *
+     * Get a list of the {@link Graph} instances and their binding names
+     *
+     * @return a {@link Map} where the key is the name of the {@link Graph} and the value is the {@link Graph} itself
      */
-    public GraphManager(final Settings settings) {
-        settings.graphs.entrySet().forEach(e -> {
-            try {
-                final Graph newGraph = GraphFactory.open(e.getValue());
-                graphs.put(e.getKey(), newGraph);
-                logger.info("Graph [{}] was successfully configured via [{}].", e.getKey(), e.getValue());
-            } catch (RuntimeException re) {
-                logger.warn(String.format("Graph [%s] configured at [%s] could not be instantiated and will not be available in Gremlin Server.  GraphFactory message: %s",
-                        e.getKey(), e.getValue(), re.getMessage()), re);
-                if (re.getCause() != null) logger.debug("GraphFactory exception", re.getCause());
-            }
-        });
-    }
+    @Deprecated
+    public Map<String, Graph> getGraphs();
 
     /**
-     * Get a list of the {@link Graph} instances and their binding names as defined in the Gremlin Server
-     * configuration file.
+     * Get a {@link Set} of {@link String} graphNames corresponding to names stored in the graph's
+     * reference tracker.
+     */
+    public Set<String> getGraphNames();
+    /**
+     * Get {@link Graph} instance whose name matches {@link gName}
      *
-     * @return a {@link Map} where the key is the name of the {@link Graph} and the value is the {@link Graph} itself
+     * @return {@link Graph} if exists, else null
+     */
+    public Graph getGraph(final String gName);
+
+    /**
+     * Add {@link Graph} g with name {@link String} gName to
+     * {@link Map<String, Graph>}
      */
-    public Map<String, Graph> getGraphs() {
-        return graphs;
-    }
+    public void addGraph(final String gName, final Graph g);
 
     /**
-     * Get a list of the {@link TraversalSource} instances and their binding names as defined by Gremlin Server
-     * initialization scripts.
+     * Get a list of the {@link TraversalSource} instances and their binding names
      *
      * @return a {@link Map} where the key is the name of the {@link TraversalSource} and the value is the
      *         {@link TraversalSource} itself
      */
-    public Map<String, TraversalSource> getTraversalSources() {
-        return traversalSources;
-    }
+    public Map<String, TraversalSource> getTraversalSources();
```

Comment:
It occurred to me while implementing `GraphManager` that `TraversalSource` objects should also be managed by the interface. At a minimum, that would mean similar methods to what we have for `Graph`.